README.md
# NAME

Plack::App::EventSource - EventSource/SSE for Plack

# SYNOPSIS

    use Plack::App::EventSource;
    use Plack::Builder;

    builder {
        mount '/events' => Plack::App::EventSource->new(
            handler_cb => sub {
                my ($conn, $env) = @_;

                $conn->push('foo');
                # or
                # $conn->push('foo', 'bar', 'baz');
                # or
                # $conn->push({id => 1, data => 'foo'});
                $conn->close;
            }
        )->to_app;

        mount '/' => $app;
    };

# DESCRIPTION

Plack::App::EventSource is an EventSource or Server Sent Events applications.
EventSource is an alternative to WebSockets when there is no need for duplex
communication. EventSource uses HTTP and is much simpler in implementation.
Ideal for website notifications or read only update streams.

It is recommended to use EventSource with polyfills to enable them in browsers
that don't support SSE. This does not need any server changes, which is very
handy. Take a look at
[EventSource.js](https://github.com/remy/polyfills/blob/master/EventSource.js)
and [jquery.eventsource](https://github.com/rwaldron/jquery.eventsource).

This library stays event loop agnostic, which means that you can use it with
[AnyEvent](https://metacpan.org/pod/AnyEvent) or [POE](https://metacpan.org/pod/POE) or even just with a plain forking server.

## Options

- `handler_cb`

    The main application entry point. It is called with
    [Plack::App::EventSource::Connection](https://metacpan.org/pod/Plack::App::EventSource::Connection) and `$env` parameters.

        handler_cb => sub {
            my ($conn, $env) = @_;

            $conn->push('hi');
            $conn->close;
        }

- `headers`

    Additional response headers. This is useful when you want to add Access Control
    headers:

        headers => [
            'Access-Control-Allow-Origin' : 'http://localhost:5000',
            'Access-Control-Allow-Credentials' : 'true'
        ]

# HOWTOs

## Sending cookies to another domain

Set CORS headers:

    headers => [
        'Access-Control-Allow-Origin' : 'http://original-domain',
        'Access-Control-Allow-Credentials' : 'true'
    ]

If you try to set `Origin` to `*` some browsers will complain. So make sure to
set the correct original domain.

When connecting to EventSource on the client side pass `withCredentials`
option:

    var es = new EventSource("http://another-domain", {
        withCredentials: true
    });

## Nginx proxy

    location /events {
            proxy_pass http://backend;
            proxy_buffering off;
            proxy_cache off;
            proxy_set_header Host $host;
            proxy_set_header Connection '';
            proxy_http_version 1.1;
            chunked_transfer_encoding off;
    }

# ISA

[Plack::Component](https://metacpan.org/pod/Plack::Component)

# METHODS

## `call($env)`

# INHERITED METHODS

## `new`

## `mk_accessors`

## `prepare_app`

## `response_cb($res, $cb)`

## `to_app`

## `to_app_auto`

# AUTHOR

Viacheslav Tykhanovskyi, <viacheslav.t@gmail.com>

# COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.


Changes
Revision history for Perl extension Plack-App-EventSource

{{$NEXT}}

0.03 2017-04-20T22:22:45Z

    - capture writer exceptions

0.02 2015-04-01T08:32:00Z

    - support all field names and documentation fixes (Jakob Voss)

0.01 2015-02-06T11:14:04Z

    - original version



lib/Plack/App/EventSource/Connection.pm
package Plack::App::EventSource::Connection;

use strict;
use warnings;

sub new {
    my $class = shift;
    my (%params) = @_;

    my $self = {};
    bless $self, $class;

    $self->{push_cb}  = $params{push_cb};
    $self->{close_cb} = $params{close_cb};

    return $self;
}

sub push {
    my $self = shift;

    $self->{push_cb}->(@_);
}

sub close {
    my $self = shift;

    $self->{close_cb}->();
}

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Plack::App::EventSource::Connection - Connection object

=head1 SYNOPSIS

Used internally.

=head1 DESCRIPTION

This is a connection object that you get in C<handler_cb> callback.

=head1 METHODS

=head2 C<new>

Creates new object.

=head2 C<close>

Closes connection.

=head2 C<push>

Pushes data to the client. Accepts an array of messages, which themselves can
be strings or hash references.

    $conn->push('message');
    $conn->push('multi', 'line', 'message');
    $conn->push({id => 1, data => 'message with id'});

possible hash fields are C<event>, C<id>, C<data>, and C<retry>.

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut


lib/Plack/App/EventSource.pm
package Plack::App::EventSource;
use 5.008001;
use strict;
use warnings;

use parent 'Plack::Component';

our $VERSION = "0.02";

use Plack::Util::Accessor qw(handler_cb headers);
use Plack::App::EventSource::Connection;

sub call {
    my $self = shift;
    my ($env) = @_;

    return [405, [], ['Method not allowed']]
      unless $env->{REQUEST_METHOD} eq 'GET';

    return sub {
        my $respond = shift;

        my $writer = $respond->(
            [
                200,
                [
                    'Content-Type' => 'text/event-stream; charset=UTF-8',
                    'Cache-Control' =>
                      'no-store, no-cache, must-revalidate, max-age=0',
                    'Access-Control-Allow-Methods' => 'GET',
                    @{$self->headers || []}
                ]
            ]
        );

        my $connection; $connection = Plack::App::EventSource::Connection->new(
            push_cb => sub {
                my (@messages) = @_;

                foreach my $msg (@messages) {
                    if (ref $msg eq 'HASH') {
                        my $event = join "\x0d\x0a",
                            map { "$_: ".$msg->{$_} }
                            grep { defined $msg->{$_} }
                            qw(event id data retry);
                        eval { $writer->write("$event\x0d\x0a"); 1 }
                          or do { $connection->close; return };
                    }
                    else {
                        eval { $writer->write("data: $msg\x0d\x0a"); 1 } or do { $connection->close; return };
                    }
                }

                eval { $writer->write("\x0d\x0a"); 1 } or do {
                    $connection->close;
                };
            },
            close_cb => sub {
                eval { $writer->close };
            }
        );

        $self->{handler_cb}->($connection, $env);
    };
}

1;
__END__
=pod

=encoding utf-8

=head1 NAME

Plack::App::EventSource - EventSource/SSE for Plack

=head1 SYNOPSIS

    use Plack::App::EventSource;
    use Plack::Builder;

    builder {
        mount '/events' => Plack::App::EventSource->new(
            handler_cb => sub {
                my ($conn, $env) = @_;

                $conn->push('foo');
                # or
                # $conn->push('foo', 'bar', 'baz');
                # or
                # $conn->push({id => 1, data => 'foo'});
                $conn->close;
            }
        )->to_app;

        mount '/' => $app;
    };

=head1 DESCRIPTION

Plack::App::EventSource is an EventSource or Server Sent Events applications.
L<EventSource|http://www.w3.org/TR/eventsource/> is an alternative to
WebSockets when there is no need for duplex communication. EventSource uses
HTTP and is much simpler in implementation.  Ideal for website notifications or
read only update streams.

This library stays event loop agnostic, which means that you can use it with
L<AnyEvent> or L<POE> or even just with a plain forking server.

Plack::App::Eventsource is a subclass of L<Plack::Component>, inheriting all
its methods. You should only have to know about the C<to_app> method.

=head2 Options

=over

=item C<handler_cb>

The main application entry point. It is called with
L<Plack::App::EventSource::Connection> and L<PSGI> C<$env> parameters.

    handler_cb => sub {
        my ($conn, $env) = @_;

        $conn->push('hi');
        $conn->close;
    }

=item C<headers>

Additional response headers. This is useful when you want to add Access Control
headers:

    headers => [
        'Access-Control-Allow-Origin' : 'http://localhost:5000',
        'Access-Control-Allow-Credentials' : 'true'
    ]

=back

=head1 HOWTOs

=head2 Client support

It is recommended to use EventSource with polyfills to enable them in browsers
that don't support SSE. This does not need any server changes, which is very
handy. Take a look at
L<EventSource.js|https://github.com/remy/polyfills/blob/master/EventSource.js>
and L<jquery.eventsource|https://github.com/rwaldron/jquery.eventsource>.

=head2 Sending cookies to another domain

Set CORS headers:

    headers => [
        'Access-Control-Allow-Origin' : 'http://original-domain',
        'Access-Control-Allow-Credentials' : 'true'
    ]

If you try to set C<Origin> to C<*> some browsers will complain. So make sure to
set the correct original domain.

When connecting to EventSource on the client side pass C<withCredentials>
option:

    var es = new EventSource("http://another-domain", {
        withCredentials: true
    });

=head2 Nginx proxy

    location /events {
            proxy_pass http://backend;
            proxy_buffering off;
            proxy_cache off;
            proxy_set_header Host $host;
            proxy_set_header Connection '';
            proxy_http_version 1.1;
            chunked_transfer_encoding off;
    }

=head1 CREDITS

Jakob Voss (nichtich)

=head1 AUTHOR

Viacheslav Tykhanovskyi, E<lt>viacheslav.t@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015-, Viacheslav Tykhanovskyi

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

This program is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for
a particular purpose.

=cut
